service/udp_test.go

```diff
@@ -140,26 +147,52 @@ func TestIPFilter(t *testing.T) {
 			},
 			payload: ciphertext,
 		}
-
-		service.GracefulStop()
-		return metrics
 	}
 
+	service.GracefulStop()
+	return metrics
+}
+
+func TestIPFilter(t *testing.T) {
+	// Test both the first-packet and subsequent-packet cases.
+	payloads := [][]byte{[]byte("payload1"), []byte("payload2")}
+
 	t.Run("Localhost allowed", func(t *testing.T) {
-		metrics := checkLocalhost(allowAll)
-		if metrics.natEntriesAdded != 1 {
-			t.Errorf("Expected 1 NAT entry, not %d", metrics.natEntriesAdded)
-		}
+		metrics := sendToDiscard(payloads, allowAll)
+		assert.Equal(t, metrics.natEntriesAdded, 1, "Expected 1 NAT entry, not %d", metrics.natEntriesAdded)
 	})
 
 	t.Run("Localhost not allowed", func(t *testing.T) {
-		metrics := checkLocalhost(onet.RequirePublicIP)
-		if metrics.natEntriesAdded != 0 {
-			t.Error("Unexpected NAT entry on rejected packet")
+		metrics := sendToDiscard(payloads, onet.RequirePublicIP)
+		assert.Equal(t, 0, metrics.natEntriesAdded, "Unexpected NAT entry on rejected packet")
```

Comment:
FYI, assert doesn't stop the execution. For that you may want `require.Equal()`.

Reply:
I don't think we need to stop execution if this fails.